linkerd/concurrency-limit/src/lib.rs

```diff
@@ -89,45 +94,42 @@ impl<T> ConcurrencyLimit<T> {
 impl<S, Request> Service<Request> for ConcurrencyLimit<S>
 where
     S: Service<Request>,
-    S::Error: Into<Error>,
 {
     type Response = S::Response;
-    type Error = Error;
+    type Error = S::Error;
     type Future = ResponseFuture<S::Future>;
 
-    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        trace!(available = %self.limit.semaphore.available_permits(), "acquiring permit");
-        try_ready!(self
-            .limit
-            .permit
-            .poll_acquire(&self.limit.semaphore)
-            .map_err(Error::from));
-
-        self.inner.poll_ready().map_err(Into::into)
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        trace!(available = %self.semaphore.available_permits(), "acquiring permit");
+        loop {
+            self.state = match self.state {
+                State::Ready(_) => {
+                    trace!(available = %self.semaphore.available_permits(), "permit acquired");
+                    return Poll::Ready(Ok(()));
+                }
+                State::Waiting(ref mut fut) => {
+                    tokio::pin!(fut);
+                    let permit = futures::ready!(fut.poll(cx));
+                    State::Ready(permit)
+                }
+                State::Empty => State::Waiting(Box::pin(self.semaphore.clone().acquire_owned())),
```

Comment:
I'm curious about this boxing...  Are we unable to name the type returned by `acquire_owned`?

Reply:
It's an `async fn`, so, yes, we are unable to name it except as a `dyn Future` trait object. Also, `Box::pin` ensures that the future is pinned, so that we can poll it while we _ourselves_ are not pinned (which `Service::poll_ready` does not guarantee).